filip/clients/ngsi_v2/iota.py

```diff
@@ -461,6 +477,131 @@ def delete_device(self, *, device_id: str,
+        import json
+        live_settings = live_device.dict(include=settings_dict)
+        new_settings = device.dict(include=settings_dict)
+
+        if not live_settings == new_settings:
+            self.delete_device(device_id=device.device_id,
+                               delete_entity=True, force_entity_deletion=True)
+            self.post_device(device=device)
+            return
+
+        # We are at a state where the device exists, but only attributes were
+        # changed.
+        # we need to update the device, and the context entry separately,
+        # as update device only takes over a part of the changes to the
+        # ContextBroker.
+
+        # update device
+        self.update_device(device=device)
+
+        # update context entry
+        # 1. build context entity from information in device
+        # 2. patch it
+        from filip.models.ngsi_v2.context import \
+            ContextEntity, NamedContextAttribute
+
+        def build_context_entity_from_device(device: Device) -> ContextEntity:
+            from filip.models.base import DataType
+            entity = ContextEntity(id=device.entity_name,
+                                   type=device.entity_type)
+
+            for command in device.commands:
+                entity.add_attributes([
+                    NamedContextAttribute(
+                        name=f"{command.name}",
+                        type=DataType.COMMAND
+                    ),
+                    NamedContextAttribute(
+                        name=f"{command.name}_info",
+                        type=DataType.COMMAND_RESULT
+                    ),
+                    NamedContextAttribute(
+                        name=f"{command.name}_status",
+                        type=DataType.COMMAND_STATUS
+                    )
+                ])
+            for attribute in device.attributes:
+                # todo metadata
+                entity.add_attributes([
+                    NamedContextAttribute(
+                        name=f"{attribute.name}",
+                        type=DataType.STRUCTUREDVALUE
+                    )
+                ])
+            for static_attribute in device.static_attributes:
+                # todo metadata
+                entity.add_attributes([
+                    NamedContextAttribute(
+                        name=f"{static_attribute.name}",
+                        type=static_attribute.type,
+                        value=static_attribute.value
+                    )
+                ])
+            return entity
+
+        if patch_entity:
+            from filip.clients.ngsi_v2 import ContextBrokerClient
+            with ContextBrokerClient(
+                    url=cb_url,
+                    fiware_header=self.fiware_headers) as client:
+
+                client.patch_entity(
+                    entity=build_context_entity_from_device(device))
+
+    def does_device_exists(self, device_id: str) -> bool:
```

Comment:
please correlate with the context broker implementation. Does it actually need to be a public method? All other users could simple use get_device and check the response

Reply:
I added it exactly for all other users, as in my opinion it is a good addition to method space. As I see it it has two advantages:

1. The user directly sees that the functionality exists, and does not work itself into it to discover that get_device throws an error if it does not exist.
2. It makes the user code cleaner, instead off always needing to use a try catch block.

The ContextBroker got a correlating method : does_entity_exists(..).
I could change it in my code to use try and catch's directly in the methods, but with the reasons above I would argument to keep these two methods.